src/Proto3/Suite/DotProto/Internal.hs

```diff
@@ -508,13 +512,34 @@ prefixedMethodName serviceName (x : xs)
       method <- typeLikeName (x : xs)
       return (fieldLikeName serviceName ++ method)
 
+prefixedMethodNameWithFlag :: MonadError CompileError m => IsPrefixed -> String -> String -> m String
+prefixedMethodNameWithFlag _ _ "" = invalidTypeNameError "<empty name>"
+prefixedMethodNameWithFlag (IsPrefixed flag) serviceName (x : xs)
+  | flag = prefixedMethodName serviceName (x : xs)
+  | isLower x = return (fieldLikeName (x : xs))
+  | otherwise = fieldLikeName <$> typeLikeName (x : xs)
+
 -- | @'prefixedFieldName' prefix field@ constructs a Haskell record selector name by prepending @prefix@ in camel-case
 -- to the message field/service method name @field@.
 prefixedFieldName :: MonadError CompileError m => String -> String -> m String
 prefixedFieldName msgName fieldName = do
   field <- typeLikeName fieldName
   return (fieldLikeName msgName ++ field)
 
+prefixedFieldNameWithFlag :: MonadError CompileError m => IsPrefixed -> String -> String -> m String
+prefixedFieldNameWithFlag (IsPrefixed flag) msgName fieldName = do
+  if flag then prefixedFieldName msgName fieldName else return $ if name `elem` keywords then name ++ "_" else toCamelCase fieldName
+  where
+    name = fieldLikeName fieldName
+    -- copy from https://hackage.haskell.org/package/hscolour-1.20.3/docs/src/Language-Haskell-HsColour-Classify.html#keywords
+    -- and remove "forall", "qualified", "ccall", "as", "safe", "unsafe"
```

Comment:
Could you mention why those keywords are removed from this check?

Also, I would rather not copy the list because there is a license mismatch.  How about an equivalent predicate, such as a case statement on the first letter, followed by `elem` on a smaller list of keywords with that first letter?  That would be faster anyway.

Reply:
Yes, but if you want to go with a Set that's fine too.

But to clarify my first question: why are you allowing "forall", "qualified", "ccall", "as", "safe", and "unsafe"?

Reply:
Thanks.